### PR TITLE
Add SRO deploy ART bundle playbook

### PR DESCRIPTION
**Pull request by pacevedom**
Added:
- New entrypoint for use from [openshift/release](https://github.com/openshift/release).
- New toolbox target for art bundle deploy.
- Playbook for art bundle deploy.

### CODE REVIEW

**Comment by kpouget** [comment]
thanks for the PR @pacevedom, with only a quick glance it looks good :) I'll look at it further tomorrow

what I would suggest is that you add a `TMP` commit on top of that to trigger `test_custom_image "$@"` from a `/test` command (eg, reroute it from `run sro test_master_branch`) so that we can see it running on the CI before reviewing/merging. (See what I do [there](https://github.com/openshift-psap/ci-artifacts/pull/235/files#diff-be3998ea816e6541eef74bb0c8a8f35a14dee47081844464bdc61faef9b3f785L282) if that's unclear)

we introduced a `/test custom` for [this purpose](https://github.com/openshift-psap/ci-artifacts/blob/master/build/root/usr/local/bin/ci_entrypoint_custom.sh) with @omertuc,  but it seems that something in Prow side doesn't work as expected :|

**Comment by pacevedom** [comment]
@kpouget thanks for the heads up! Didnt know that technique, I will include it in the next commit. I am still missing the most difficult part though, which is to get the ART images out to another registry.

File: roles/sro_deploy_custom_image/tasks/deploy.yml
Comment (kpouget) [inline roles/sro_deploy_custom_image/tasks/deploy.yml:22]: can you write these logs to a file, eg `{{ artifact_extra_logs_dir }}/sro_controller_manager.log`?

overall, I try to keep the task output concise, and store everything that can be useful for troubleshooting and/or understanding what happened during the execution inside files in `{{ artifact_extra_logs_dir }}`

File: toolbox/special_resource_operator.py
Comment (kpouget) [inline toolbox/special_resource_operator.py:38]: by reading the content of this PR, it's not really clear what this image is. 
I think it will be an image for the SRO operator/manager Pod, built by ART,
but I think you can clarify a bit here and there

File: build/root/usr/local/bin/ci_entrypoint_sro.sh
Comment (kpouget) [inline build/root/usr/local/bin/ci_entrypoint_sro.sh:39]: I think @omertuc woud usually suggests to use something like that:
```
    CUSTOM_IMAGE="$1"
    shift || true
    CI_IMAGE_SRO_COMMIT_CI_REPO="${1:-https://github.com/openshift-psap/special-resource-operator.git}"
    shift || true
    CI_IMAGE_SRO_COMMIT_CI_REF="${1:-master}"
```

**Review by kpouget — CHANGES_REQUESTED** [review]
thanks @pacevedom, looks good overall :)
I pointed out a few nits, nothing serious

File: roles/sro_deploy_custom_image/meta/main.yml
Comment (pacevedom) [reply roles/sro_deploy_custom_image/meta/main.yml]: Done

File: roles/sro_deploy_custom_image/tasks/deploy.yml
Comment (pacevedom) [reply roles/sro_deploy_custom_image/tasks/deploy.yml]: Done

File: toolbox/special_resource_operator.py
Comment (pacevedom) [reply toolbox/special_resource_operator.py]: The commit looks different now, so I renamed it to be more clear.

File: roles/sro_deploy_custom_image/tasks/deploy.yml
Comment (pacevedom) [reply roles/sro_deploy_custom_image/tasks/deploy.yml]: This file is different now, but there is extra logs dir in some key places.

File: build/root/usr/local/bin/ci_entrypoint_sro.sh
Comment (pacevedom) [reply build/root/usr/local/bin/ci_entrypoint_sro.sh]: I kept the file's style, but I can do this change if needed.

Comment (kpouget) [reply build/root/usr/local/bin/ci_entrypoint_sro.sh]: yes, we usually try to focus on having *new* code following our informally-defined common best practise, so if you want to update this block as I suggested ... and perhaps the rest of the file for consistency :)

Comment (pacevedom) [reply build/root/usr/local/bin/ci_entrypoint_sro.sh]: Done in e2a143808b08741d5d5604d3f8524c6dd547205f

File: roles/sro_deploy_art_bundle/files/builder_namespace.yml
Comment (kpouget) [inline roles/sro_deploy_art_bundle/files/builder_namespace.yml:7]: new line at end of file

(I'll add a Github Action to enforce that soon ;-))

File: roles/sro_deploy_art_bundle/tasks/main.yml
Comment (kpouget) [inline roles/sro_deploy_art_bundle/tasks/main.yml:5]: this is a `shell`, not a `command`
and you'll need `set -o pipefail` to avoid (legit) `ansible-lint` complains

can you split long commands into multiple lines?
for this one I would suggest something like:
```
shell: 
  oc get secret/pull-secret -n openshift-config -o json 
      | jq -r '.data.".dockerconfigjson"' 
      | base64 -d 
      > "{{ artifact_extra_logs_dir }}/authfile"
```

File: roles/sro_deploy_art_bundle/tasks/main.yml
Comment (kpouget) [inline roles/sro_deploy_art_bundle/tasks/main.yml:11]:
```
command:
    oc set data secret/pull-secret 
       -n openshift-config 
       --from-file=.dockerconfigjson="{{ artifact_extra_logs_dir }}/authfile"
```

File: roles/sro_deploy_art_bundle/tasks/main.yml
Comment (kpouget) [inline roles/sro_deploy_art_bundle/tasks/main.yml:22]: that's a `shell` with `pipefail`

File: roles/sro_deploy_art_bundle/tasks/main.yml
Comment (kpouget) [inline roles/sro_deploy_art_bundle/tasks/main.yml:176]: not `clusterpolicy`, maybe `CR`

File: roles/sro_deploy_art_bundle/templates/sro_sub.yml.j2
Comment (kpouget) [inline roles/sro_deploy_art_bundle/templates/sro_sub.yml.j2:8]: not sure you want to keep `channel: "{{ sro_channel }}"` commented out

**Review by kpouget — COMMENTED** [review]
looks good @pacevedom, I pointed out a few nits, mainly `command` --> `shell` and long lines that can but reformatted to ease the reading

I couldn't spot any big red flag in the Ansible code, but obviously it's not tested yet, so running it live will certainly highlights other things ;-)

we'll merge https://github.com/openshift-psap/ci-artifacts/pull/254 and especially https://github.com/openshift-psap/ci-artifacts/pull/255 sooner or later, so maybe you can rebase on top of #255 and see how to get your PR tested

File: roles/sro_deploy_art_bundle/files/builder_namespace.yml
Comment (pacevedom) [reply roles/sro_deploy_art_bundle/files/builder_namespace.yml:7]: Done

File: roles/sro_deploy_art_bundle/tasks/main.yml
Comment (pacevedom) [reply roles/sro_deploy_art_bundle/tasks/main.yml:5]: Done

File: roles/sro_deploy_art_bundle/tasks/main.yml
Comment (pacevedom) [reply roles/sro_deploy_art_bundle/tasks/main.yml:11]: Done

File: roles/sro_deploy_art_bundle/tasks/main.yml
Comment (pacevedom) [reply roles/sro_deploy_art_bundle/tasks/main.yml:22]: Done

File: roles/sro_deploy_art_bundle/tasks/main.yml
Comment (pacevedom) [reply roles/sro_deploy_art_bundle/tasks/main.yml:176]: Done

File: roles/sro_deploy_art_bundle/templates/sro_sub.yml.j2
Comment (pacevedom) [reply roles/sro_deploy_art_bundle/templates/sro_sub.yml.j2:8]: That was for testing... it slipped.

File: roles/sro_deploy_art_bundle/tasks/main.yml
Comment (kpouget) [inline roles/sro_deploy_art_bundle/tasks/main.yml:9]: @pacevedom you shouldn't store `authfile` and `dockerconfigjson` files in the `extra_logs_dir`, both of them have secret content!! 
you can thank the testplatform team for doing the proper censoring that prevented the leak here ;-)

https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift-psap_ci-artifacts/247/pull-ci-openshift-psap-ci-artifacts-master-sro-e2e-master-art/1438537855172349952/artifacts/e2e-master-art/nightly/artifacts/006__sro__deploy_art_bundle/authfile
https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift-psap_ci-artifacts/247/pull-ci-openshift-psap-ci-artifacts-master-sro-e2e-master-art/1438537855172349952/artifacts/e2e-master-art/nightly/artifacts/006__sro__deploy_art_bundle/dockerconfigjson

I guess you can write into `/tmp` *and* delete the files in a `always` block!

**Comment by pacevedom** [comment]
Waiting on https://github.com/openshift-psap/special-resource-operator/pull/95 to test this again.

**Comment by kpouget** [comment]
@pacevedom 

if you have a bit of time to give it a try, would be nice to update to remove this from [the test logs](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift-psap_ci-artifacts/247/pull-ci-openshift-psap-ci-artifacts-master-sro-e2e-master-art/1442876890057543680/artifacts/e2e-master-art/nightly/artifacts/006__sro__deploy_art_bundle/_ansible.log):

```
2021-09-28 15:55:11,355 p=2788 u=psap-ci-runner n=ansible | ---
2021-09-28 15:55:11,355 p=2788 u=psap-ci-runner n=ansible | roles/sro_deploy_art_bundle/tasks/main.yml:49
2021-09-28 15:55:11,356 p=2788 u=psap-ci-runner n=ansible | TASK: sro_deploy_art_bundle : Remove git clone
2021-09-28 15:55:11,356 p=2788 u=psap-ci-runner n=ansible |   - path:	/tmp/special-resource-operator
2021-09-28 15:55:11,356 p=2788 u=psap-ci-runner n=ansible |   - diff
2021-09-28 15:55:11,356 p=2788 u=psap-ci-runner n=ansible |     - before
2021-09-28 15:55:11,356 p=2788 u=psap-ci-runner n=ansible |       - path:	/tmp/special-resource-operator
2021-09-28 15:55:11,356 p=2788 u=psap-ci-runner n=ansible |       - state:	directory
2021-09-28 15:55:11,356 p=2788 u=psap-ci-runner n=ansible |       - path_content
2021-09-28 15:55:11,357 p=2788 u=psap-ci-runner n=ansible |         - directories:	['/tmp/special-resource-operator/.git',
# goes on forever with the list of all the files in the directory -,-
```

I think [this map](https://github.com/openshift-psap/ci-artifacts/blob/master/callback_plugins/human_log.py#L9) should be updated:
```
INTERESTING_MODULE_PROPS = {
    "stat": ["path", "exists", "mode"],
    "invocation": None,
    "_ansible_no_log": None,
    "_ansible_verbose_always": None,
    "__ansible_verbose_override": None,
    "msg": None,
    "changed": None,
    "ansible_facts": None,
    "include_args": None,
    "failed_when_result": None,
    }
```
with something like
```
"diff": None,
```
to fully ignore the diff block from the logging.
That's piece of code isn't the cleanest ever, but that's the best I could come up with to turn Ansible logs into something nice to read :)

**Comment by kpouget** [comment]
image

would you be able to generate an `operator.version` file and store it in `ARTIFACTS_DIR`?

see what I do for the [GPU Operator](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/periodic-ci-openshift-psap-ci-artifacts-release-4.9-gpu-operator-e2e-master/1442987710514991104/artifacts/gpu-operator-e2e-master/nightly/artifacts/operator.version):
```
9.9.9-latest.git.3ace7c4
```
which gets displayed in the dashboard:
image
(I will change the GPU Operator master-branch version to `<commit date>.git.<commit short hash>` in the next PR)

File: roles/sro_deploy_art_bundle/tasks/main.yml
Comment (kpouget) [inline roles/sro_deploy_art_bundle/tasks/main.yml:163]: I'm not sure this should be considered as a flake, as you're building the catalog yourself?

File: roles/sro_deploy_art_bundle/tasks/main.yml
Comment (kpouget) [inline roles/sro_deploy_art_bundle/tasks/main.yml:248]: nit: empty line between tasks

File: toolbox/special_resource_operator.py
Comment (kpouget) [inline toolbox/special_resource_operator.py:38]: can you add a basic description for `git_repo` and `git_ref` arguments?

File: roles/sro_deploy_art_bundle/files/builder_script.yml
Comment (pacevedom) [reply roles/sro_deploy_art_bundle/files/builder_script.yml:41]: Cant use the same auth file for podman push because thats an internal registry using service account credentials. I placed the file existing check, though.

File: roles/sro_deploy_art_bundle/files/builder_script.yml
Comment (kpouget) [reply roles/sro_deploy_art_bundle/files/builder_script.yml:12]: you should be able mute the error with something like
```
match=$(egrep -m 1 ".+\.${commit}\..+" ${BUNDLES} || true)
```

File: roles/sro_deploy_art_bundle/files/builder_script.yml
Comment (kpouget) [reply roles/sro_deploy_art_bundle/files/builder_script.yml:41]: what I meant (just a suggestion) is:
1. pod spec:
```
   - mountPath: /var/lib/kubelet/config.json
      name: node-pullsecrets
      readOnly: true

    env:
      - name: NODE_PULLSECRET
        value: /var/lib/kubelet/config.json
```
2 . `podman`
```
BUNDLES="/tmp/bundles"
podman search --authfile $NODE_PULLSECRET
```
3. `opm`
```
REGISTRY_AUTH_FILE=$NODE_PULLSECRET /usr/local/go/bin/opm index add --bundles ${BUNDLE} --tag ${CATALOG_INDEX} -c podman
```

or just call the env `REGISTRY_AUTH_FILE`, instead of `NODE_PULLSECRET`
anyway, it's mostly bikeshedding, do as you prefer :)

File: roles/sro_deploy_art_bundle/tasks/main.yml
Comment (pacevedom) [reply roles/sro_deploy_art_bundle/tasks/main.yml:163]: Swapped it with describe+logs.

File: roles/sro_deploy_art_bundle/files/builder_script.yml
Comment (pacevedom) [reply roles/sro_deploy_art_bundle/files/builder_script.yml:46]: The bundle I get from ART is already built, so everything that is left is to include it in an index and then spawn a pod from it. I chose not to use quay because I didnt want the index to leave the cluster, so I turned to the local image registry. This required some RBAC policies for the openshift-marketplace namespace and the new catalog pod service account. Had no trouble apart from that, which prevents the pod from starting.

**Comment by kpouget** [comment]
@pacevedom any update on this PR? still ongoing work, or blocked, or obsoleted?

**Comment by pacevedom** [comment]
> @pacevedom any update on this PR? still ongoing work, or blocked, or obsoleted?

Ready for review, I think it is still valuable to introduce it.

File: roles/sro_deploy_art_bundle/templates/sro_sub.yml.j2
Comment (kpouget) [reply roles/sro_deploy_art_bundle/templates/sro_sub.yml.j2:8]: still there

File: roles/sro_deploy_art_bundle/tasks/main.yml
Comment (kpouget) [inline roles/sro_deploy_art_bundle/tasks/main.yml:126]: can you turn this into an `always` block, so that the logs and description are always stored (mostly logs, description is less interesting when things go right)

File: roles/sro_deploy_art_bundle/tasks/main.yml
Comment (kpouget) [inline roles/sro_deploy_art_bundle/tasks/main.yml:182]: same, if you can turn this into `always`, to follow what's happening from the artifacts

**Review by kpouget — CHANGES_REQUESTED** [review]
thanks @pacevedom, looks good, I just pointed out a few nits,
should be ready for merging once addressed :)

File: roles/sro_deploy_art_bundle/tasks/main.yml
Comment (pacevedom) [reply roles/sro_deploy_art_bundle/tasks/main.yml:182]: Done

File: roles/sro_deploy_art_bundle/tasks/main.yml
Comment (pacevedom) [reply roles/sro_deploy_art_bundle/tasks/main.yml]: Done

**Comment by kpouget** [comment]
Hello @pacevedom , PR looks good to me, only this:

image

but you marked it as resolved, so you're happy with it?

**Comment by pacevedom** [comment]
> Hello @pacevedom , PR looks good to me, only this:
> 
> image
> 
> but you marked it as resolved, so you're happy with it?

My bad, slipped again. Should be fixed now.

**Comment by kpouget** [comment]
good for me @pacevedom :)

/lgtm
/approve
/hold
I let you unhold if you're ready for merging

**Review by kpouget — APPROVED** [review]
good for me, marking the PR as "approved" on Github interface

File: roles/sro_deploy_art_bundle/files/builder_buildconfig.yml
Comment (kpouget) [inline roles/sro_deploy_art_bundle/files/builder_buildconfig.yml:31]: @pacevedom, any reason for building `opm` here, instead of [downloading](https://github.com/operator-framework/operator-registry/releases/download/v1.19.5/linux-amd64-opm) it? (as @omertuc noticed)

**Comment by kpouget** [comment]
@pacevedom can we close this PR?

**Comment by pacevedom** [comment]
Closing because of switching to different type of testing.

**Pull request closed without merging**
